docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,424 @@
+
+## Non-Goals
+
+* In the current scope, capability to abort/suspend an `out-of-band` task is not going to be provided. This could be considered as an enhancement based on pull.
+* Ordering (by establishing dependency) of `out-of-band` tasks submitted for the same etcd cluster has not been considered in the first increment. In a future version based on how operator tasks are used, we will enhance this proposal and the implementation.
+
+## Proposal
+
+Authors propose creation of a new single dedicated custom resource to represent an `out-of-band` task. Etcd-druid will be enhanced to process the task requests and update its status which can then be tracked/observed.
+
+### Custom Resource API
+
+`EtcdOperatorTask` is the new custom resource that will be introduced. This API will be in `v1alpha1` version and will be subject to change. We will be respecting [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).
+
+```go
+// EtcdOperatorTask represents an out-of-band operator task resource.
+type EtcdOperatorTask struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // Spec is the specification of the EtcdOperatorTask resource.
+  Spec EtcdOperatorTaskSpec `json:"spec"`
+  // Status is most recently observed status of the EtcdOperatorTask resource.
+  Status EtcdOperatorTaskStatus `json:"status,omitempty"`
+}
+```
+
+#### Spec
+
+The authors propose that the following fields should be specified in the spec (desired state) of the `EtcdOperatorTask` custom resource.
+
+* To capture the type of `out-of-band` operator task to be performed, `.spec.type` field should be defined. It can have values from all supported `out-of-band` tasks eg. "OnDemandSnaphotTask", "QuorumLossRecoveryTask" etc.
+* To capture the configuration specific to each task, a `.spec.config` field should be defined of type `string` as each task can have different input configuration.
+
+```go
+// EtcdOperatorTaskSpec is the spec for a EtcdOperatorTask resource.
+type EtcdOperatorTaskSpec struct {
+  
+  // Type specifies the type of out-of-band operator task to be performed. 
+  Type string `json:"type"`
+
+  // Config is a task specific configuration.
+  Config string `json:"config,omitempty"`
+
+  // TTLSecondsAfterFinished is the time-to-live to garbage collect the 
+  // related resource(s) of task once it has been completed.
+  // +optional
+  TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+
+  // OwnerEtcdReference refers to the name and namespace of the corresponding 
+  // Etcd owner for which the task has been invoked.
+  OwnerEtcdRefrence types.NamespacedName `json:"ownerEtcdRefrence"`
+}
+```
+
+#### Status
+
+The authors propose that the following fields should be specified in the `Status` (current state) of the `EtcdOperatorTask` custom resource as the custom resource's `Status` will be used to monitor the progress of the task.
```

Comment:
For `Status` using `should be specified` is incorrect, since the users never set anything in `Status`. I am in general not very happy about only describing a few fields and leaving out the rest. For `Spec` it can still be fine because you have defined `mandatory` fields but there also its really not required. I am of the opinion that you either describe/detail out every field or leave out all and only make it as part of the doc-string in the golang snippet.